lib/rspec/expectations/expectation_target.rb

```diff
@@ -90,6 +90,45 @@ def prevent_operator_matchers(verb)
       include InstanceMethods
     end
 
+    # @private
+    # Validates the provided matcher to ensure it supports block
+    # expectations, in order to avoid user confusion when they
+    # use a block thinking the expectation will be on the return
+    # value of the block rather than the block itself.
+    class ValueExpectationTarget < ExpectationTarget
+      def to(matcher=nil, message=nil, &block)
+        enforce_value_expectation(matcher)
+        super
+      end
+
+      def not_to(matcher=nil, message=nil, &block)
+        enforce_value_expectation(matcher)
+        super
+      end
+
+    private
+
+      def enforce_value_expectation(matcher)
+        return if supports_value_expectations?(matcher)
+
+        RSpec.deprecate(
+          "expect(value).to #{RSpec::Support::ObjectFormatter.format(matcher)}",
+          :message =>
+            "You must pass a block rather than an argument to `expect` to use the provided " \
```

Comment:
Does this message need rewording? It should probably talk about it being recommended rather than must.

Reply:
> I don't recall how it was done in 2.99, don't we fail on deprecation warnings in the pre-major version?
No. Or they wouldn't be warnings.

Reply:
Rather than `"You must"` rephrase to `"<x> is deprecated, and you should do <y> instead"`

Reply:
The message is fixed.